docs/basics/101-136-history.rst

```diff
@@ -0,0 +1,767 @@
+
+.. runrecord:: _examples/DL-101-136-129
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ datalad status
+
+.. runrecord:: _examples/DL-101-136-130
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ datalad save -m "add joke evaluation to joke" Gitjoke2.txt
+
+How could this modification to ``Gitjoke2.txt`` be undone?
+With the :command:`git reset` command again. If you want to
+"unsave" the modification but keep it in the file, use
+:command:`git reset --mixed` as before. However, if you want to
+get rid of the modifications entirely, use the option ``--hard``
+instead of ``--mixed``:
+
+.. runrecord:: _examples/DL-101-136-131
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ git log -2 --oneline
+
+.. runrecord:: _examples/DL-101-136-132
+   :language: console
+   :workdir: dl-101/DataLad-101
+   :realcommand: echo "$ git reset --hard $(git rev-parse HEAD~1)" && git reset --hard $(git rev-parse HEAD~1)
+
+.. runrecord:: _examples/DL-101-136-133
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ cat Gitjoke2.txt
+
+The change has been undone completely. This method will work with
+files stored in Git and annexed files.
+
+Undoing past modifications of files
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+What :command:`git reset` did was to undo commits from
+the most recent version of your dataset. How
+would one undo a change that happened a while ago, though,
+with important changes being added afterwards that you want
+to keep?
+
+Let's save a bad modification to ``Gitjoke2.txt``,
+but also a modification to ``notes.txt``:
+
+.. runrecord:: _examples/DL-101-136-140
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ echo "bad modification" >> Gitjoke2.txt
+
+.. runrecord:: _examples/DL-101-136-141
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ datalad save -m "did a bad modification" Gitjoke2.txt
+
+.. runrecord:: _examples/DL-101-136-142
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ cat << EOT > notes.txt
+
+   Git has many handy tools to go back in forth in
+   time and work with the history of datasets.
+   Among many other things you can rewrite commit
+   messages, undo changes, or look at previous versions
+   of datasets. A superb resource to find out more about
+   this and practice such Git operations is this
+   chapter in the Pro-git book:
+   https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
+
+   EOT
+
+.. runrecord:: _examples/DL-101-136-143
+   :language: console
+   :workdir: dl-101/DataLad-101
+
+   $ datalad save -m "add note on helpful git resource" notes.txt
+
+The objective is to remove the first, "bad" modification, but
+keep the more recent modification of ``notes.txt``. A :command:`git reset`
+command is not convenient, because resetting would need to reset
+the most recent, "good" modification as well. Instead, an interactive
+rebase with the command :command:`git rebase -i` is necessary [#f5]_.
```

Comment:
The presented approach is valid, but also problematic in any collaborative scenario where the respective branch was pushed already, and others may have already built on on. Rebasing in this case leads to a world of pain that quickly escalates and requires everyone involved to have some expert level Git foo.

The most compatible, least intrusive way revert one or more changes is `git revert` IMHO. It will produce (an) additional commit(s) (instead of rewriting history). This makes it obvious in the history that a mistake happened that was rectified. Any consumer can simply merge this change and also sees that a mistake has been fixes -- in contrast to a rather obscure merge conflict.

Most of the text on conflict resolution below remains valid in the context of `git revert` and can and should be kept.

Reply:
Done.